libtenzir/builtins/operators/flatten.cpp

```diff
@@ -67,12 +67,40 @@ class flatten_operator final : public crtp_operator<flatten_operator> {
   std::string separator_ = default_flatten_separator;
 };
 
-class plugin final : public virtual operator_plugin<flatten_operator> {
+class plugin final : public virtual operator_plugin<flatten_operator>,
+                     public virtual function_plugin {
 public:
   auto signature() const -> operator_signature override {
     return {.transformation = true};
   }
 
+  auto make_function(invocation inv,
+                     session ctx) const -> failure_or<function_ptr> override {
+    auto expr = ast::expression{};
+    auto sep = std::optional<std::string>{default_flatten_separator};
+    TRY(argument_parser2::function("tql2.flatten")
+          .add(expr, "<field>")
+          .add(sep, "<string>")
+          .parse(inv, ctx));
+    return function_use::make([expr = std::move(expr),
+                               sep = std::move(sep.value())](
+                                evaluator eval, session ctx) -> series {
+      auto s = eval(expr);
+      auto seen = std::unordered_set<type>{};
+      auto flattened = tenzir::flatten(
+        s.type, std::static_pointer_cast<arrow::StructArray>(s.array), sep);
+      // We only warn once per schema that we had to rename a set of fields.
+      if (seen.insert(s.type).second && not flattened.renamed_fields.empty()) {
+        diagnostic::warning("renamed fields with conflicting names after "
+                            "flattening: {}",
+                            fmt::join(flattened.renamed_fields, ", "))
+          .primary(expr)
+          .emit(ctx);
+      }
```

Comment:
This deduplication (A) does not work, because `seen` is re-constructed every time the function is run, but also (B) is not needed, because we already do warning deduplication. Just make sure that the primary message is the same if you want to deduplicate it.